### PR TITLE
[Model] Update Paligemma multimodal processing with PromptUpdate 

### PR DESCRIPTION
**Pull request by kylehh**
Continue work from #13584
1. Update MM processing with PromptUpdate 
2. Fixed the issue of can not identify language with "caption \<lang\>" prompt 

Note
1. Use PromptReplacement for paligemma and PromptInsertion for Paligemma 2
2. The issue was due to HF encode prompt adding \<bos\> before \<image\> tokens (ref [here](https://github.com/huggingface/transformers/blob/25245ec26dc29bcf6102e1b4ddd0dfd02e720cf5/src/transformers/models/paligemma/processing_paligemma.py#L161)) > . The fix was to remove \<bos\> for baseline result to pass text_common test

### CODE REVIEW

File: vllm/model_executor/models/paligemma.py [hunk 1]
```diff
@@ -46,95 +49,171 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
...
-        hidden_states = self.linear(image_features)
-        return hidden_states
+        # Remove <bos> token at the beginning of prompt added by HF processor
+        # Otherwise it will fail the language feature
+        # This is for Paligemma 1 model only (tokenizier.add_bos_token == True)
+        # Paligemma2 does NOT have this problem (add_bos_token == False)
+        if processed_outputs["input_ids"][0][0] == tokenizer.bos_token_id:
+            prompt_ids_without_bos = processed_outputs["input_ids"][0][1:]
+            processed_outputs["input_ids"] = prompt_ids_without_bos[None, :]
+        return processed_outputs
```
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:140]: I think we don't need this anymore since we now use `PromptInsertion`?

Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: This is to deal w <bos> at very beginning. Need to remove it for language feature to work.

Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: And this code is to process hf procssor results, so it can match with prompt replacemtn result

Comment (DarkLight1337) [reply vllm/model_executor/models/paligemma.py:140]: Where does the bos token come from? Does it come from the HF processor itself?

Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: Yes, it's from [HF processor ](https://github.com/huggingface/transformers/blob/25245ec26dc29bcf6102e1b4ddd0dfd02e720cf5/src/transformers/models/paligemma/processing_paligemma.py#L265). The `input_string` is "\<image\>...\<image\>\<bos\>Initial str" and the call of tokenizer would add the \<bos\> at the very beginning due to add_bos_token setting is True for paligemma 1

**Comment by DarkLight1337** [comment]
Please also merge the latest changes from #14038

File: vllm/model_executor/models/paligemma.py (hunk 1 above)
Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: Yes it would output correctly but only in English, even with lang set to es. That was my previous implementation. This only affect paligemma 1 though since we dont have this issue w paligemma 2

Comment (DarkLight1337) [reply vllm/model_executor/models/paligemma.py:140]: Hmm, that shouldn't happen. Can you show how you ran the HF implementation of the model?

Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: I ran `pytest test_common.py` and the `baseline_output` is the one calling HF processor, and the `processed_outputs = super()._call_hf_processor` from PaliGemmaMultiModalProcessor:_call_hf_processor should be the directly output from HF, right?

Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:140]: Because my test shows paligemma 1 can not get language correct ( with main vLLM branch, using HF processor results) and paligemma 2 can. I compared the two and I think it's maybe a bug in HF implementation so modified in vllm to make paligemma 1 and 2 both have same format and now both version can output language correctly.

File: vllm/model_executor/models/paligemma.py [hunk 2]
```diff
@@ -46,95 +49,171 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
...
-        self.linear = nn.Linear(vision_hidden_size, projection_dim, bias=True)
+    def _call_hf_processor(
+        self,
+        prompt: str,
+        mm_data: Mapping[str, object],
+        mm_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        tokenizer = self.info.get_tokenizer()
+        if not mm_data:
+            # HF processor always adds placeholders even when there's no image
```
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:123]: We can remove this comment as it's not relevant.

File: vllm/model_executor/models/paligemma.py (hunk 1 above)
Comment (kylehh) [reply vllm/model_executor/models/paligemma.py]: OK. Reverting back to HF processor without modification.

File: vllm/model_executor/models/paligemma.py [hunk 3]
```diff
@@ -46,95 +49,161 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
...
+    def _call_hf_processor(
+        self,
+        prompt: str,
+        mm_data: Mapping[str, object],
+        mm_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        tokenizer = self.info.get_tokenizer()
+        if not mm_data:
+            prompt_ids = tokenizer.encode(prompt)
+            return BatchFeature(dict(input_ids=[prompt_ids]), tensor_type="pt")
```
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:124]: Can the HF processor be called on text-only inputs? If so, we don't have to override this method.

File: vllm/model_executor/models/paligemma.py [hunk 4]
```diff
@@ -46,95 +49,161 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
...
+            return [
+                PromptInsertion(
+                    modality="image",
+                    target=[],
+                    insertion=PromptUpdateDetails(
+                        full=image_tokens + [bos_token_id],
+                        features=image_tokens,
+                    ),
+                )
+            ]
```
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:179]:
```suggestion
        # Paligemma 1 and 2 have different tokenizer.add_bos_token
        # Replace <bos> with <bos> + <image>*n + <bos> for Paligemma 1
        # Insert <image>*n + <bos> for Paligemma 2
        return [
            PromptReplacement(
                modality="image",
                target=PromptIndexTargets.prefix([bos_token_id] if tokenizer.add_bos_token else []),
                replacement=PromptUpdateDetails(
                    full=image_tokens + [bos_token_id],
                    features=image_tokens,
                ),
            )
        ]
```

Simplify things a bit

File: vllm/model_executor/models/paligemma.py (hunk 3 above)
Comment (kylehh) [reply vllm/model_executor/models/paligemma.py:124]: I tested but job failed without this code to deal with text only input

Comment (DarkLight1337) [reply vllm/model_executor/models/paligemma.py:124]: Alright, let's keep this then.

**Comment by DarkLight1337** [comment]
It looks like multi-modal tests is still failing. To get around the accuracy issue, feel free to switch to PaliGemma 2.

**Comment by DarkLight1337** [comment]
> It looks like multi-modal tests is still failing. To get around the accuracy issue, feel free to switch to PaliGemma 2.

We can still keep PaliGemma 1 in `test_common.py`, just not in the model correctness tests.

**Review by DarkLight1337 — APPROVED** [review]
I have updated the docs for you. Thanks for your hard work!
